### PR TITLE
Fix RTP in nether

### LINKED ISSUE #13: RTP teleports the player to the overworld when run in other dimensions
RTP teleports the player to the overworld when run in other dimensions. I expected it to either fail or teleport the player within the current dimension.

### Versions
- ftb-essentials-1604.1.0.13

### PR DESCRIPTION
**Pull request by mkg20001**
This adds code partly copied from PortalForcer (vanilla) to spiral arround the location and search for a good spot
if the position returned by heightmap turns out to be above world height (such as one on top of
the bedrock layer of the nether)

This is better than teleporting arround in another dimension

Fixes #13

### CODE REVIEW

File: src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java [hunk 1]
```diff
@@ -89,10 +94,15 @@ public static int rtp(ServerPlayer player) {
 		FTBEPlayerData data = FTBEPlayerData.get(player);
 		return data.rtpTeleporter.teleport(player, p -> {
 			p.displayClientMessage(new TextComponent("Looking for random location..."), false);
-			return findBlockPos(p.server.getLevel(Level.OVERWORLD), p, 1);
+			return findBlockPos(player.getLevel(), p, 1);
 		}).runCommand(player);
 	}
 
+	private static boolean isBlockPosAir(ServerLevel world, BlockPos pos) {
+		BlockState bs = world.getBlockState(pos);
```
Comment (LatvianModder) [inline src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: This should just be `world.getBlockState(pos).isAir()`

File: src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java [hunk 2]
```diff
@@ -125,11 +135,24 @@ private static TeleportPos findBlockPos(ServerLevel world, ServerPlayer player,
...
+		BlockPos hmPos = world.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, currentPos);
+
+		if (hmPos.getY() > 0) {
+			if (hmPos.getY() >= world.getHeight()) { // broken heightmap (nether, other mod dimensions)
+				Iterator<BlockPos.MutableBlockPos> it = BlockPos.spiralAround(new BlockPos(hmPos.getX(), world.getSeaLevel(), hmPos.getY()), 16, Direction.EAST, Direction.SOUTH).iterator();
+				while (it.hasNext()) {
+					BlockPos newPos = it.next();
+					BlockState bs = world.getBlockState(newPos);
+
+					if (bs.getMaterial().isSolidBlocking() && !bs.is(BlockTags.LEAVES) && !bs.is(Blocks.BEDROCK) && isBlockPosAir(world, newPos.above(1)) && isBlockPosAir(world, newPos.above(2)) && isBlockPosAir(world, newPos.above(3))) {
```
Comment (LatvianModder) [inline src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: Hm, this should use `ftbessentials:ignore_rtp` block tag that includes these two

**Review by LatvianModder — DISMISSED** [review]

**Comment by LatvianModder** [comment]
Are you sure that this wont end up spawning players above bedrock? Its possible for mushrooms to grow on top of it, so could that maybe mess with the bedrock check?

**Comment by mkg20001** [comment]
It checks if it's above world.getHeight() and then uses the workarround method to find a better location deeper down.

As the mushroom isn't solid it shouldn't be possible to spawn ontop of the world.

File: src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java (hunk 2 above)
Comment (mkg20001) [reply src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: I can't find ignore_rtp tag, where is it defined?

File: src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java (hunk 1 above)
Comment (mkg20001) [reply src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: blockstate .isAir() is deprecated, used .getBlock().isAir(world, state, pos) instead

File: src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java (hunk 2 above)
Comment (MaxNeedsSnacks) [reply src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: You should define the tag yourself (statically, using `TagHooks.getBlockOptional`) and just add the leaves tag and bedrock as values to it by default ^^ This enables server admins to configure custom blocks they don't want players to be rtp'd on top of

Comment (mkg20001) [reply src/main/java/dev/ftb/mods/ftbessentials/command/TeleportCommands.java]: I'm really not the person to throw tasks at. It wasn't there previously aswell. Maintainer push is enabled on this PR.

**Review by MaxNeedsSnacks — APPROVED** [review]

**Pull request merged**